### PR TITLE
Detect object properties usage inside getUsages

### PR DESCRIPTION
**Pull request by academo**
Closes: https://github.com/grafana/levitate/issues/149

Adds detection of access to object properties to the `getUsage` API. 

This doesn't affect (yet) the is-compatible command as the compareDiff doesn't report object properties .

Simple case:

```typescript
        import { FooBar } from 'testing-module';
        const object = new FooBar();
        object.method(); // detect that method was used as part of FooBar an import of "testing-module";

```

More complex case

```typescript
        import { FooBar } from 'testing-module';
        const Proxy = FooBar;
        const t1 = new Proxy();
        const t2 = t1;
        t2.method(); // detect that method was used as part of FooBar an import of "testing-module";
```

More examples of detection can be found in the unit tests.

### CODE REVIEW

File: fixtures/usage/object-properties/package.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
```
Comment (academo) [inline fixtures/usage/object-properties/package.json:1]: This package.json is required for the fixtures we use for testing and development.

Typescript can only report the correct parent symbols if they are declared and loaded as modules from node_modules.

File: src/bin.ts [hunk 2]
```diff
@@ -220,4 +221,12 @@ yargs
     console.log('Try running levitate with --help to see available commands.');
     exit(1);
   })
-  .help().argv;
+  .help();
+
+// in DEBUG mode this allows the debugger to connect and disconnect more easily
+if (process.env.DEBUG) {
```
Comment (academo) [inline src/bin.ts:227]: While working with the nodejs I noticed the debugger gets into a deadlock because of how yargs handle the exit events that I can't overwrite. 

This is a solution to force the debut to disconnect when the process emits the exit event.

Notice that usually is useful to have the debugger still connected when the script finishes but in this specific case the debugger is not usable and the only thing you can do with it is to close it or refresh it.

Comment (leventebalogh) [inline src/bin.ts:227]: **Suggestion:** I think we can get rid of the condition here (`if (process.env.DEBUG) { ... }`) as it is already checked in the underlying function.

File: src/utils/typescript.ts [hunk 3]
```diff
@@ -22,6 +22,17 @@ export function getAllIdentifiers(node: ts.Node | ts.SourceFile): ts.Identifier[
   return identifiers;
 }
 
+export function getAllPropertyAccessExpressions(node: ts.Node | ts.SourceFile): ts.PropertyAccessExpression[] {
+  const identifiers: ts.PropertyAccessExpression[] = [];
```
Comment (leventebalogh) [inline src/utils/typescript.ts:26]: **Suggestion:** I would rather use the name `expressions` for this variable here (for consistency)
```suggestion
  const expresions: ts.PropertyAccessExpression[] = [];
```

File: src/utils/typescript.ts [hunk 4]
```diff
@@ -36,8 +47,11 @@ export function getSymbolFromParameter(node: ts.ParameterDeclaration, program: t
   return undefined;
 }
 
-export function createTsProgram(fileName: string): ts.Program {
-  const program = ts.createProgram([fileName], COMPILER_OPTIONS);
+export function createTsProgram(fileName: string, compilerOptions: ts.CompilerOptions = {}): ts.Program {
+  const program = ts.createProgram([fileName], {
+    ...COMPILER_OPTIONS,
+    ...compilerOptions,
+  });
```
Comment (leventebalogh) [inline src/utils/typescript.ts:54]: **Suggestion:** I would either rename `COMPILER_OPTIONS` to `COMPILER_OPTIONS_DEFAULT`, or `compilerOptions` to `compilerOptionsOverrides`.

File: fixtures/usage/object-properties/use-theme.tsx [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+const getStyles = (theme: GrafanaTheme2) => ({
+  styleA: css`
+    padding: theme.spacing(3);
+    ${theme.breakpoints.up('md')} {
+      display: flex;
+    }
+  `,
+});
+
+export default MyComponent;
```
Comment (leventebalogh) [inline fixtures/usage/object-properties/use-theme.tsx:23]: **Question:** are we using these fixtures anywhere?

**Review by leventebalogh — CHANGES_REQUESTED** [review]
Nice new feature! I also really like that we have tests covering it. 👏

File: src/bin.ts [hunk 6]
```diff
@@ -220,4 +221,12 @@ yargs
...
+  forceDebugExit();
+}
```
[the remark this answers is not in the document]
Comment (academo) [reply src/bin.ts:229]: Ahh! that was because of a different approach I had before. I can put it back now that I did this way.

File: src/bin.ts (hunk 2 above)
Comment (academo) [reply src/bin.ts:227]: Maybe. But I want to make emphasis in the code that this is a debug-only function.

File: src/usage/usage.ts [hunk 7]
```diff
@@ -73,5 +74,28 @@ function getUsageOfSourceFile(
       usage[identifierName].count = (usage[identifierName].count || 0) + 1;
     }
   }
+
+  const checker = project.getTypeChecker();
```
[the remark this answers is not in the document]
Comment (academo) [reply src/usage/usage.ts:78]: For now. But I am sure we will need more than the checker later on with modifications. Also it makes no difference.

File: fixtures/usage/object-properties/use-theme.tsx (hunk 5 above)
Comment (academo) [reply fixtures/usage/object-properties/use-theme.tsx:23]: Yes. When I develop levitate I use them extensively.

File: src/usage/test-utils.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
+import { getExportInfo } from '../compiler/exports';
+import { generateTmpFileWithContent } from '../tests/test-utils';
+import { IdentifierWithCounter } from '../types';
+import { createTsProgram } from '../utils/typescript';
+import { getPackageUsage } from './usage';
+
+export function getUsageForTesting({
+  projectSrc,
+  testingModuleSrc,
+  testingModuleName = 'testing-module',
```
[the remark this answers is not in the document]
Comment (academo) [reply src/usage/test-utils.ts:10]: I don't really think this makes a difference.

File: src/usage/usage.test.ts [hunk 9]
```diff
@@ -24,57 +24,55 @@ describe('Usage', () => {
...
     Qux();
     `;
-    const projectFile = generateTmpFileWithContent(projectSrc);
 
-    const testingModuleExports = getExportInfo(testingModuleFile);
+      const use = getUsageForTesting({
+        projectSrc,
+        testingModuleSrc: testingModuleSrc,
+        testingModuleName: 'testing-module',
+      });
```
[the remark this answers is not in the document]
Comment (academo) [reply src/usage/usage.test.ts:43]: A valid concern but I don't think it hides anything. The name of the function and parameters clearly indicate what it does and the way is used makes it clearer once you see the code.

I don't see any advantage of having more repetitive code in these tests.

File: src/bin.ts (hunk 2 above)
Comment (leventebalogh) [reply src/bin.ts]: Hehe, let's add the condition one more time then, so it's easier to understand.

Jokes aside, we can also leave the condition in this file only and remove it from the function. Or keep it in the function and change its name so it reveals that it should be used for debug purposes. It's not crucial from the _functional_ perspective, but the current implementation is redundant and redundancies make the code harder to reason about later. Maybe a line today, but a year later a thousand more little redundancies like this take us to a codebase which is harder to read and understand.

File: src/usage/usage.test.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (mckn) [reply src/usage/usage.test.ts:43]: I agree with Levi here. I think the tests serve two purposes.

1. To verify that the functionality works as expected now and in the future when we introducing changes.
2. Documentation so when any of us (or a new person) re-visit the code base, they can read the tests to understand the functionality and how it is supposed to be used.

I had a quick look at the code since we are abstracting away the setup in multiple files you need to navigate 3 different files to understand how it is setup. With that said I think we can make some small improvements to make the tests more readable. 

Also, a nit from my side. The way we currently have formatted the template strings makes it really hard to read. So changing that would be a quick win!

File: src/usage/test-utils.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (jackw) [reply src/usage/test-utils.ts:10]: Maybe it's just me but I try to name arguments the same as a functions parameters where possible (which I think is what Levi is suggesting here). I certainly find it easier to skim read code as things are passed around and I have to go from file to file.

File: src/usage/usage.test.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (academo) [reply src/usage/usage.test.ts:43]: @mckn can you elaborate on the template strings formatting? What makes it difficult to read?

The testing code will become very repetitive if we split it. I don't really think much or any is achieved by changing it but if the consensus is I should I'll do the change.

[the remark this answers is not in the document]
Comment (academo) [reply src/usage/usage.test.ts:43]: I pushed the change to remove some responsability from the test utility back to the test itself.

File: src/usage/usage.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (academo) [reply src/usage/usage.ts:78]: To further expand my previous comment:

While working on other parts of leviate I identified that the program and checker are far more important to have at hand than only the typescript nodes or symbols and you can't reliably fetch the program from the symbols.

In this specific case we only need the checker but I suspect, as we have to expand the functionality to cover more cases like destructuring, import aliases and other edge cases, we'll need more of the utilities and properties included in both the program and the checker.

In other words I learnt that it is better to have the program at hand at any moment you are doing operations over nodes and symbols.

File: src/usage/test-utils.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (academo) [reply src/usage/test-utils.ts]: I renamed them to use the word "package" as suggested.

File: src/usage/usage.test.ts [hunk 10]
```diff
@@ -1,168 +1,395 @@
...
-    const testingModuleExports = getExportInfo(testingModuleFile);
-    const otherModuleExports = getExportInfo(otherModuleFile);
+      const { projectProgram, testingModuleExports, testingPackageName, projectFile } = getProjectUsageSetupForTesting({
+        projectSrc,
+        testingPackageSrc,
+        testingPackageName: 'testing-module',
+      });
+      const usages = getPackageUsage(projectProgram, testingModuleExports, testingPackageName);
+      const mainFile = Array.from(usages.keys()).find((f) => f.fileName === projectFile);
+      const use = usages.get(mainFile);
```
Comment (leventebalogh) [inline src/usage/usage.test.ts:97]: Would it make sense to extract these lines to a utility function in this file to remove the repetition?
```javascript
const getUsedNamesForFile = (usages: Map<ts.SourceFile, Record<string, IdentifierWithCounter>>, fileName: string) => {
   const file = Array.from(usages.keys()).find((f) => f.fileName === filename);

    return usages.get(file);
}

// ...

const allUsages = getPackageUsage(projectProgram, testingModuleExports, testingPackageName);
const usedNames = getUsedNamesForFile(allUsages, projectFile);

expect(usedNames).toEqual(['Bar']);
```

File: src/bin.ts (hunk 2 above)
Comment (leventebalogh) [reply src/bin.ts]: Looking at it from another angle, what is the reason for having it in both places?

File: src/usage/usage.test.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (leventebalogh) [reply src/usage/usage.test.ts:43]: > The testing code will become very repetitive if we split it. I don't really think much or any is achieved by changing it but if the consensus is I should I'll do the change.

I am sorry if I was not clear on this, I didn't mean that we should make the code more repetitive, more like that if we use the `getPackageUsage()` function explicitly in the tests then they are easier to be used as a documentation as well. What do you think about https://github.com/grafana/levitate/pull/176#discussion_r977519815 to reduce some of the repetition?

[the remark this answers is not in the document]
Comment (mckn) [reply src/usage/usage.test.ts:43]: > @mckn can you elaborate on the template strings formatting? What makes it difficult to read?
> 
> The testing code will become very repetitive if we split it. I don't really think much or any is achieved by changing it but if the consensus is I should I'll do the change.

As I wrote in PM. It is the code snippets within e.g. `projectSrc` which you already have fixed 🙌🏻

**Review by leventebalogh — APPROVED** [review]
